[WIP]: Add GID-shifting support for future dual-network-simulation

asoplata opened the pull request:
Turning this into a PR just to keep track of the conversation, even if we end up shifting some of the work into another branch.

Myrausman committed:
feat: add dual network simulation script for HNN-Core

Myrausman committed:
Merge branch 'jonescompneurolab:master' into two-network-simulation

Myrausman committed:
Merge branch 'jonescompneurolab:master' into two-network-simulation

Myrausman committed:
Merge branch 'jonescompneurolab:master' into two-network-simulation

Myrausman committed:
debugging

Myrausman committed:
Merge branch 'jonescompneurolab:master' into two-network-simulation

Myrausman committed:
gid_list_fix

Myrausman committed:
Refactor connection name generation in NetworkBuilder and add minimal dual network simulation script

Myrausman committed:
Merge branch 'jonescompneurolab:master' into testing_minimal_dual_network

Myrausman committed:
undoing changes for gids made earlier in other branch

Myrausman committed:
Merge branch 'testing_minimal_dual_network' of https://github.com/Myrausman/hnn-core into testing_minimal_dual_network

Myrausman committed:
udate plotting functions for improved flexibility

Myrausman committed:
Add gid_start parameter to network functions for flexible gid assignment

asoplata commented on hnn_core/network.py:
```diff
@@ -1492,10 +1495,14 @@ def add_tonic_bias(
         """Add cell type by updating pos_dict and gid_ranges."""
-        ll = self._n_gids
-        self._n_gids += len(pos)
+        if gid_start is None:
```
This could be a one-liner since the logic is simple, like
```
ll = self._n_gids if (gid_start is None) else gid_start
```

Myrausman committed:
Add debug statements to network and network_builder for improved traceability

Myrausman committed:
Remove debug statements from NetworkBuilder class for cleaner code

Myrausman committed:
type changed from string to int for src_gid, targe_gid

Myrausman committed:
clean up debug statements and improve formatting

Myrausman committed:
Add gid_start parameter to add_evoked_drive

Myrausman committed:
Refactor gid_pairs handling in create_minimal_network

Myrausman committed:
clean up extra loc

asoplata committed:
Make drives identical, add dipole plotting

asoplata committed:
swap hardcoded celltype names in dipole calc

Myrausman committed:
Add suffix parameter to Network class and update dipole cell types in simulation

Myrausman committed:
Add gid_start parameter to drive addition methods in Network class

Myrausman committed:
Add dipole_cell_types parameter to Network class and update NetworkBuilder references

Myrausman committed:
Add debug statements for combined network simulation

Myrausman committed:
Merge branch 'master' into testing_minimal_dual_network

Myrausman committed:
Refactor dipole data handling in parallel backends and remove unused network combination functions

Myrausman committed:
Merge branch 'jonescompneurolab:master' into testing_minimal_dual_network

Myrausman committed:
remove unused imports

Myrausman committed:
Add get_next_gid method to Network class and update simulation to use it

Myrausman committed:
Enhance tests for minimal dual network: add helper functions for dipole handling and improve drive configuration tests

asoplata commented on hnn_core/network.py:
```diff
@@ -1535,10 +1547,11 @@ def add_tonic_bias(
                 )
 
-    def _add_cell_type(self, cell_name, pos, cell_template=None):
+    def _add_cell_type(self, cell_name, pos, cell_template=None, gid_start=None):
```
Please add tests inside `tests/test_network.py` for most (ideally all) your changes that you have made to both `network.py` AND `network_builder.py`. Specifically, `network_builder.py` does not have its own test file currently, and instead its tests should go in `test_network.py`.

asoplata commented on hnn_core/tests/test_minimal_dual_network.py:
```diff
@@ -0,0 +1,271 @@
+import os
```
The `minimal_dual_network_simulation.py` script does not need tests. Since it is an example or tutorial of *how* to do the dual network simulations, it doesn't need tests, but should be converted into an "example/tutorial" kind of format. I wouldn't worry about adding too much example documentation to `minimal_dual_network_simulation.py` for now, however, since the example script itself is a WIP based on the future "true" combined network feature.

Myrausman committed:
Remove debug print statements from dipole and network classes

Myrausman committed:
Add tests for cell type addition and dipole behavior validation

Myrausman committed:
resolve cnflicts

Myrausman committed:
Add example for simulating two independent minimal networks

asoplata committed:
Merge branch 'master' into testing_minimal_dual_network

asoplata committed:
maint: clean conflicts and temporary test removal

asoplata committed:
ruff format

asoplata committed:
maint: very minor cleanup/reverting

asoplata committed:
maint: remove dipole_cell_types arg

`Network`'s new `dipole_cell_types` is no longer needed, due to upstream
changes on `master`, which now enable celltypes to have metadata
including a `measure_dipole` flag.

asoplata committed:
maint: revert Dipole data expansion

It is too early to merge a change like this in, for at least the
following reasons:

1. There is still incomplete work (and bugs) preventing the current
state of this branch from successfully running a true "combined" dual
network simulation. There would need to be other work done before this
code is actually used, even for testing. That code is better suited to a
later PR instead of this one.

2. This only works for the case where there are exactly two celltypes
contributing to the dipole in each layer, instead of dynamically
organizing the data to fit multiple sizes.

3. Instead of hard-coding data-assignments in `dipole.py`, it would be
better to eventually (in a later PR) use the brand-new celltype
metadata (see here:
https://github.com/jonescompneurolab/hnn-core/blob/6b099a7dfe3f3ddbb3cd8c1472832089d80158bf/hnn_core/network_models.py#L84
) to organize the data shape.

Therefore, I will be removing expanded data shape code.

asoplata committed:
descope/revert: revert suffix/naming code

Originally, this example
script (`examples/howto/plot_dual_network_simulation.py`) made two
distinct networks which were different in two ways:

1. `net2`'s first GID was the first *after* the last GID of `net1` (with
or without `net1`'s drive), and

2. The celltypes in `net2` were renamed so that they all ended with a
new suffix, `_net2`.

These changes were made so that, *eventually*, in the future, when other
bugs were fixed and work was done, this same script would *then*
eventually be extended so that a user would simulate BOTH `net1` and
`net2` *simultaneously, as part of the same complete Network*. The
GID-reassignment code would be necessary to have the two Networks not
collide in GID space, and the celltype suffixes would be necessary to
differentiate between the appropriate celltype of `net1` versus `net2`.

However, after some investigation, I believe this PR needs to be
re-scoped to *only* include the GID-reassignment code, but *not* the
celltype-suffix code. The GID-reassignment code appears to work
currently, but the celltype-suffix does not.

Unfortunately, the celltype-suffix code is reliant on function
`Network._rename_cell_types` which exhibits the following:

1. It does NOT guarantee that dipole etc. output will be identical
"before vs. after" its use; see the following:
https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/network.py#L1599
. It appears that, in some cases, it *was* producing identical output
before vs after in some prior commits of this PR. However, that may have
been a coincidence, since more work is needed to actually ensure that
`Network._rename_cell_types` is robust to Dipole calculation (including
testing that specifically, which would currently fail).

2. More importantly, the failure of `Network._rename_cell_types` to
produce dipole-identical output has been expanded by something in the
following recent (and large) commit:
https://github.com/jonescompneurolab/hnn-core/commit/6388f9f7c74f4943a1d9b06e9ae69314500913df
. In other words, Maira's dual-independent-networks scripts *before*
this commit do produce identical output in `net1` and `net2`. However,
after this commit, something has changed with either
`Network._rename_cell_types` or its dependencies, and the same
scripts *after* this commit do *not* produce identical output between
`net1` and `net2`.

As a result of this, I see two possible options to move forward in this
PR:

1. I fix `Network._rename_cell_types` to produce Dipole-identical
output, *including* while dealing with Chetan's new changes in the above
commit. Then, this PR includes both options for GID-reassignment and
that change. *I don't think this is a good idea, and would increase the
time-to-merge and scope significantly*.

2. I shrink the scope of this PR to exclude and remove any kind of
celltype-naming code, and make this PR exclusively about adding
GID-reassignment code. *I believe this is the best way forward*.

There was always going to be some clashing of Chetan's GSoC
main-objective of de-hardcoding celltypes, versus the minimum
changes (of the same sort) that Maira's GSoC objective needed in order
to lay the foundation and begin her work. Since both projects are still
incomplete, but there is *far* more work remaining to do for the
dual-network simulation feature, I think that changing this PR to *only*
be about adding GID-reassignment functionality is the best way forward.

asoplata committed:
fix: round robin cell GID bug

Maira's previous changes introduced round-robin allocation of MPI per
cell GIDs, but only *per celltype*. This caused a nasty bug where, if
you were using MPI and your number of processors exceeded the max number
of cells in any cell type, the simulation would fail. This fixes that by
using a single round-robin list, across different celltypes.

asoplata committed:
refactor: move gid shift code into Network

asoplata committed:
doc: add docstring for shift

asoplata committed:
maint: privatize gid functions

Despite having a public example illustrating their use, I believe these
GID-shift functions should be currently marked as private, and NOT part
of the public API. This is because they are not intended to be used by
the general public, and the major feature (simultaneous multiple network
simulation) is still far from complete, so they are only useful in
development *for* that feature.

I have left the example how-to in with the rest of the how-tos, but we
may want to consider moving it, since it is a useful example, but
currently is only useful for *development purposes only*.

asoplata committed:
maint/doc/fmt: force shift arg, docstrings, ruff

asoplata committed:
test: add tests for new net methods and drive args

asoplata committed:
doc: update changelog with current status

asoplata committed:
fix: forgot some privatization

dylansdaniels commented on examples/howto/plot_dual_network_simulation.py:
```diff
@@ -0,0 +1,172 @@
+print("Next GIDs after drives:", net1._get_next_available_gid(), net2._get_next_available_gid())
+
+###############################################################################
+# Step 7: (Optional) Export Configurations
```
It seems like we don't really need this optional section? There's no explanation to why we're doing this and the written outputs are never used

asoplata replied:
Deleted in latest commit.

dylansdaniels commented on examples/howto/plot_dual_network_simulation.py:
```diff
@@ -0,0 +1,172 @@
+dpl1 = simulate_dipole(net1, tstop=20.0, dt=0.025, n_trials=1)
+dpl2 = simulate_dipole(net2, tstop=20.0, dt=0.025, n_trials=1)
+
+print("Net1 dipole peak:", max(abs(dpl1[0].data['agg'])))
```
Some comments on why we're showing these things would be good

asoplata committed:
Apply Dylan suggestions

asoplata committed:
more emphasis on net1/net2 diff

asoplata commented:
Ready for round 2 of review

asoplata committed:
doc: remove docstring changes for easier review

asoplata committed:
Merge branch 'master' into testing_minimal_dual_network

asoplata committed:
doc: Nick-suggested changes

asoplata committed:
doc: add note about motivation for plot raster

asoplata commented on examples/howto/plot_dual_network_simulation.py:
```diff
@@ -0,0 +1,171 @@
+# Step 9: Combined Spike Raster
+# ------------------------------
+
+def plot_combined_spike_raster(net_a, net_b, title="Combined Spike Raster"):
```
This is kind of a validation plot, since this is still in beta: this allows us to confirm that everything is working correctly, even for a GID-shifted network versus a non-GID-shifted one (same inputs leads to same outputs). This plot specifically was what Maira was using for much of the debugging, especially including cell-type-specific discrepancies between the two networks as they arose. I wanted to leave this in here especially because it will be a useful validation function when we return to developing this functionality.

asoplata commented on examples/howto/plot_dual_network_simulation.py:
```diff
@@ -0,0 +1,171 @@
+"""
+==============================================
+10. Simulating Two Minimal Independent Networks
```
You raise a good point. Just to make things more complicated, we are in the process of moving all of our "execution of example notebooks" over to the Textbook repo here: https://github.com/jonescompneurolab/textbook . Once we do that, we are planning to *remove* the Examples from the Developer Portal website here https://jonescompneurolab.github.io/hnn-core/stable/auto_examples/index.html since the Textbook will have replaced the examples. When that happens, we will still be able to keep this example script as a notebook on the Textbook site for future use, but *without* actually publishing or showing it on the webpage unless we want to. 

In other words, after we finish updating the Textbook execution stuff, we can choose whether or not to display this example script/notebook, while still keeping it around.

ntolley commented on hnn_core/network.py:
```diff
@@ -1019,6 +1032,10 @@ def add_bursty_drive(
         conn_seed : int
             Optional initial seed for random number generator (default: 3).
             Used to randomly remove connections when probability < 1.0.
+        gid_start : int, default: None
```
Maybe add the default values for the synaptic delays and space constants like it was done in the previous modifications? No need to overhaul all the docstrings, but just to keep consistent with what is changed in the other drives

ntolley commented on hnn_core/network.py:
```diff
@@ -1583,10 +1615,25 @@ def add_tonic_bias(
+            drive starting from the Network's max GID + 1.
+
+        """
+        ll = self._n_gids if (gid_start is None) else gid_start
```
what is `ll`?, is there a more explicit variable name we could use?

ntolley commented on hnn_core/network.py:
```diff
@@ -2175,6 +2222,80 @@ def _standardize_spike_data(self, spike_data):
+        max_gid = -1
+        for rng in self.gid_ranges.values():
+            if len(rng) > 0:
+                max_gid = max(max_gid, max(rng))
```
```suggestion
        for gid_range in self.gid_ranges.values():
            if len(gid_range) > 0:
                max_gid = max(max_gid, max(gid_range))
```
Nitpick but `rng` is pretty exclusively used to mean random number generator in a lot of code

ntolley commented on hnn_core/network.py:
```diff
@@ -2175,6 +2222,80 @@ def _standardize_spike_data(self, spike_data):
+                ct = conn["src_type"]
+                gid_range = list(self.gid_ranges[ct])
+                n_srcs = len(conn["src_gids"])
+                conn["src_gids"] = gid_range[:n_srcs]
```
this code could be condensed into a loop over `['src_type', 'target_type']`

ntolley commented on hnn_core/network.py:
```diff
@@ -2175,6 +2222,80 @@ def _standardize_spike_data(self, spike_data):
+                n_srcs = len(conn["src_gids"])
+                conn["src_gids"] = gid_range[:n_srcs]
+
+            if "gid_pairs" in conn and conn["gid_pairs"]:
```
The code looks correct to me so I think this would pass, but is there an explicit test ensuring that the same cells are connected when `conn['probability'] < 1.0`?

I think this is actually one of the most critical sections of code as it ensures that the same cells are connected even after their ids change

asoplata replied:
Added some in the latest commit, including greatly expanding that the connectivity (using equal `conn_seeds`) uses the same "shape", but shifted, between the base and shifted networks: https://github.com/Myrausman/hnn-core/commit/8445b45f5f0b5810ffaf5a0dd16e0454fe1fe514

ntolley commented on hnn_core/network_builder.py:
```diff
@@ -408,7 +408,13 @@ def _gid_assign(self, rank=None, n_hosts=None):
-        for gid in range(self._rank, self.net._n_cells, n_hosts):
+        gids = list()
+        for cell_type, gid_range in self.net.gid_ranges.items():
+            # Only assign real cell types (not drives) here
```
Does MPI break if drives are added before shifting?

ntolley commented on hnn_core/tests/test_network.py:
```diff
@@ -1911,3 +1911,98 @@ def test_update_weights_metadata():
+    """Test that `Network._shift_gid_ranges` works and simulates as expected."""
+    # "Base" has many connections and drives, but for now we're only interested in the
+    # non-drive params config
+    net_base, params = base_network
```
```suggestion
    _, params = base_network
```

ntolley reviewed (changes requested):
@asoplata I gave this a full read through! I think this is close to merge, the most pressing comment is the last one on the test, I think I am misunderstanding `get_next_available_gid()`

asoplata committed:
ref: nick example scritp refactor [no ci]

asoplata committed:
doc: minor update of network docstrings [no ci]

asoplata committed:
[no ci] ref: minor nick review changes

asoplata commented on hnn_core/network.py:
```diff
@@ -2175,6 +2222,80 @@ def _standardize_spike_data(self, spike_data):
+          `Network` and additional `Networks` use a disjoint set of GIDs.
+        - This successfully shifts GIDs for individual cells and for non-drive synapses,
+          but currently does NOT shift GIDs for external drives. You should shift your
+          Network's GIDs only BEFORE adding drives to the Network.
```
Resolved in a new commit.

asoplata committed:
ref: remove connectivity checks in shift code etc.

asoplata committed:
ugh ruff

asoplata committed:
[no ci] ref: drive test to use multiple sizes

asoplata committed:
test: add shift test with prob<1.0

asoplata committed:
test/ref: split GIDshift tests for parallel [no ci]

asoplata committed:
test: expand post-GID-shift drive tests

asoplata committed:
[no ci] valid: Add error if drives b4 shift

asoplata committed:
Merge branch 'master' into testing_minimal_dual_network

asoplata committed:
doc: move up changelog addition

asoplata committed:
[MRG] bugfix: Update args to fix GUI viz tests (#1217)

* [MRG] bugfix: Update args to fix GUI viz tests

This is a tiny, trivial bugfix that fixes the currently failing tests on
`master`.

`master` currently fails
`test_gui.py::test_gui_visualization`
due to an error in the `"network"` plot_type case of the main `for`
loop.

The failing testing code is here:

```python
setup_gui = <hnn_core.gui.gui.HNNGUI object at 0x10a36f050>

    def test_gui_visualization(setup_gui):
        """Tests updating a figure creates plots with data."""

        gui = setup_gui
        # Spectrogram needs longer time for wavelet analysis
        gui.widget_tstop.value = 500
        gui.run_button.click()

        gui._simulate_viz_action("switch_fig_template", "[Blank] single figure")
        gui._simulate_viz_action("add_fig")
        figid = 2
        figname = f"Figure {figid}"
        axname = "ax0"

        for viz_type in _plot_types:
            gui._simulate_viz_action(
                "edit_figure", figname, axname, "default", viz_type, {}, "clear"
            )
            # Check that extra axes have been successfully removed
            assert len(gui.viz_manager.figs[figid].axes) == 1
            # Check if data on the axes has been successfully cleared
            assert not gui.viz_manager.figs[figid].axes[0].has_data()

            gui._simulate_viz_action(
                "edit_figure", figname, axname, "default", viz_type, {}, "plot"
            )
            # Check if data is plotted on the axes
>           assert gui.viz_manager.figs[figid].axes[0].has_data()
E           AssertionError: assert False
E            +  where False = has_data()
E            +    where has_data = <Axes: label='Spike histogram'>.has_data

hnn_core/tests/test_gui.py:718: AssertionError
```

while the actual problem is in the provided Traceback here:

```python
Traceback (most recent call last):
  File "/opt/anaconda3/envs/hc11/lib/python3.11/site-packages/ipywidgets/widgets/widget.py", line 191, in __call__
    local_value = callback(*args, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/austinsoplata/rep/brn/hnn-core/hnn_core/gui/_viz_manager.py", line 577, in _plot_on_axes
    dpls_processed = _update_ax(
                     ^^^^^^^^^^^
  File "/Users/austinsoplata/rep/brn/hnn-core/hnn_core/gui/_viz_manager.py", line 441, in _update_ax
    img_arr = np.reshape(
              ^^^^^^^^^^^
TypeError: reshape() got an unexpected keyword argument 'newshape'
```

Specifically, the root cause is that `_viz_manager.py::_update_ax()`, in
the `plot_type == "network"` case, makes a call to `np.reshape`. The
named parameter in that call is named `newshape`, however, in Numpy
2.4.0, this named arg was changed to `shape`, which you can see in the
release notes here:
https://numpy.org/doc/stable/release/2.4.0-notes.html#removed-newshape-parameter-from-numpy-reshape
Renaming the named arg allows all tests to pass once more.

Note that after merge, this change will need to be merged/rebased into
any existing PRs in order for them to pass tests.

* fix: remove named arg entirely for older versions

asoplata committed:
doc: fix whats_new conflict

asoplata committed:
test: test of prevent gid shifting when drive/bias

asoplata committed:
test: expand main GID shift test

This adds the following to the biggest and most comprehensive
GID-shifting test:

1. Adds optional use of `inter_network_spacing` to test both cases where
either
    A. the second (shifted) network's GID array follows consecutively
    after the first network's, or
    B. the second (shifted) network's GID array follows some "empty" GID
    spacing after the first network's.

2. Add a `<net>.cell_response.spike_times` comparison between the
un-shifted network and the shifted network, to check that their output
is also identical.

3. Add optional usage of MPI in this same test, so that the effects of
GID shifting are tested in both MPI and non-MPI contexts.

asoplata committed:
Merge branch 'master' into testing_minimal_dual_network

ntolley commented on examples/howto/plot_dual_network_simulation.py:
```diff
@@ -0,0 +1,171 @@
+# Step 2: Helper to Build a Minimal Network
+# -----------------------------------------
+#
+# Creates a network with one cell per canonical cell type. Their GIDs reassigned to
```
```suggestion
# Creates a network with one cell per canonical cell type. Their GIDs are then reassigned to
```

ntolley commented on examples/howto/plot_dual_network_simulation.py:
```diff
@@ -0,0 +1,171 @@
+net1.add_evoked_drive(
+    'evdist1_net1', mu=5.0, sigma=1.0, numspikes=1, location='distal',
+    weights_ampa={'L2_pyramidal': 0.1, 'L5_pyramidal': 0.1},
+    gid_start=drive_gid_1
```
Does it really make sense to expose this to the public API? If you set this value smaller than the next available gid, then there's an error (?)

But if it is larger, then there's a weird gap in the gid range

asoplata commented on hnn_core/tests/test_network.py:
```diff
@@ -2073,3 +2079,212 @@ def test_check_global_synaptic_gains_uniformity():
+    _ = simulate_dipole(net_shifted, tstop=20.0, n_trials=1)
+
+
+@pytest.mark.parametrize("probability", [1.0, 0.67])  # SIX SEVEN /dance
```
are you frame-mogging my comment???
